### PR TITLE
LetterSpacing

### PR DESCRIPTION
**Pull request by swingingtom**
This contains the implementation of LetterSpacing #98  and a example.
Im not sure the example worth to stay there, but at least you can use it as a line-break validator.

I choose to put computed `letterSpacing` value as [Text inlines.element](https://github.com/swingingtom/three-mesh-ui/blob/acfe9a07bb267b94a2304c0b4d4720aafb9a888c/src/components/Text.js#L102) property because the letterSpacing value was also required in [InlineManager.distanceToNextBreak](https://github.com/swingingtom/three-mesh-ui/blob/acfe9a07bb267b94a2304c0b4d4720aafb9a888c/src/components/core/InlineManager.js#L200). Which only had `Text.inlines` as available parameters right now. 
It waste a bit of memory because every `Text.inlines.element` of a same `Text` has now the same value behind `letterSpacing` property. But I did not want to start nor to propose a refactoring of `InlineManager.distanceToNextBreak`

This is why updating the `letterSpacing` property might also require a [parsingNeedsUpdate](https://github.com/swingingtom/three-mesh-ui/blob/acfe9a07bb267b94a2304c0b4d4720aafb9a888c/src/components/core/MeshUIComponent.js#L375)

### CODE REVIEW

File: examples/letter_spacing.js [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+	container.add(
+
+		text,
+
+		new ThreeMeshUI.Text({
+			content: " As well as multi-font-size lines with consistent vertical spacing.",
+			fontSize: 0.08
+		})
+
+	);
```
Comment (felixmariotto) [inline examples/letter_spacing.js:96]: I think the example is useful, but you should make the text more explanatory. For example you could remove the second `Text` component and update the first one's content to `"letterSpacing ".repeat(3)`.

File: examples/letter_spacing.js [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+		let letterSpacing = text.getLetterSpacing();
+		letterSpacing += letterSpacingSpeed;
+
+		// reverse the speed at some points
+		if( letterSpacing > 1 || letterSpacing <= 0 ){
+			letterSpacingSpeed *= -1;
+		}
+
+		// update the text component
+		text.set({letterSpacing:letterSpacing});
```
Comment (felixmariotto) [inline examples/letter_spacing.js:112]: This is cleaner:
`text.set( { letterSpacing } );`

File: src/components/Text.js [hunk 3]
```diff
@@ -104,7 +105,8 @@ export default class Text extends mix.withBase( Object3D )(
                 anchor: dimensions.anchor,
                 lineBreak,
                 glyph,
-                fontSize
+                fontSize,
+                letterSpacing: letterSpacing
             };
```
Comment (felixmariotto) [inline src/components/Text.js:111]: cleaner:
```
return {
    height: dimensions.height,
    width: dimensions.width,
    anchor: dimensions.anchor,
    lineBreak,
    glyph,
    fontSize,
    letterSpacing
};
```

File: src/components/core/MeshUIComponent.js [hunk 4]
```diff
@@ -266,6 +266,10 @@ export default function MeshUIComponent( Base = class {} ) {
             return (this.hiddenOverflow === undefined) ? DEFAULTS.hiddenOverflow : this.hiddenOverflow;
         }
 
+        getLetterSpacing() {
+          return (this.letterSpacing === undefined) ? DEFAULTS.letterSpacing : this.letterSpacing;
+        }
+
```
Comment (felixmariotto) [inline src/components/core/MeshUIComponent.js:272]: Here I think the property could be fetched from a parent with `_getProperty`, but I will discuss this more in length later

**Comment by felixmariotto** [comment]
Thanks a lot for your work, this is very clean.

There is things that I think could be improved with your implementation though:
- it needs a parsingUpdate (as you mentioned), so it's a bigger performance hit.
- you can't set it on the container, which would arguably be more convenient in some cases.

Firstly, in order to be able to set `letterSpacing` on a parent containing many `Text` components, you should fetch `letterSpacing` with [_getProperty](https://github.com/felixmariotto/three-mesh-ui/blob/5e3a68fb6b980fc6d72166ab1ecac502e79a2f4a/src/components/core/MeshUIComponent.js#L126) in `MeshUIComponent`.

Secondly I don't think we should need a parsing update [here](https://github.com/felixmariotto/three-mesh-ui/blob/acfe9a07bb267b94a2304c0b4d4720aafb9a888c/src/components/core/MeshUIComponent.js#L378). We should be able to pass letters spacing to `InlineManager` without having to recompute all the letters dimensions.

[This is where](https://github.com/felixmariotto/three-mesh-ui/blob/5e3a68fb6b980fc6d72166ab1ecac502e79a2f4a/src/components/core/InlineManager.js#L44) I propose to get the spacing value. You could do something like this:
`const lp = inlineComponent.isText ? inlineComponent.getLetterSpacing() : 0;`

**Comment by swingingtom** [comment]
I've update the PR with your suggestions.

I've overwritten the [second fontSize case](https://github.com/felixmariotto/three-mesh-ui/blob/5e3a68fb6b980fc6d72166ab1ecac502e79a2f4a/src/components/core/MeshUIComponent.js#L390) with letterSpacing.

I've just a question about the following : 
> This is where I propose to get the spacing value. You could do something like this:
> const ls = inlineComponent.isText ? inlineComponent.getLetterSpacing() : 0;

Is that useful only because we can put pictures and icons which don't have to suffer from letterSpacing ? Because as this property is now available on blocks and text, if this is not set, it will defaults to zero anyway.

**Comment by felixmariotto** [comment]
OK it looks good to me.

> Is that useful only because we can put pictures and icons which don't have to suffer from letterSpacing ?

Well since the property propagates to children, I thought users might be annoyed that `letterSpacing` also has an effect on `InlineBlocks` when setting it on the root component. I'm not sure about this though, it's really a design decision. What do you think ?

**Comment by swingingtom** [comment]
I lack pratices with this library. I really just have used Text component for now, so I really can't tell.
The point behind that question was to understand a bit more 'why', and your response answered that.

**Comment by felixmariotto** [comment]
Merged and published to NPM as v5.5.0, thanks a lot !

**Comment by swingingtom** [comment]
Unfortunately something went bad, and line-breaking is now wrong on all examples

**Pull request merged**